runtime/allocatable.h

```diff
@@ -0,0 +1,88 @@
+//===-- runtime/allocatable.h -----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// Defines APIs for Fortran runtime library support of code generated
+// to manipulate and query allocatable variables, dummy arguments, & components.
+#ifndef FORTRAN_RUNTIME_ALLOCATABLE_H_
+#define FORTRAN_RUNTIME_ALLOCATABLE_H_
+#include "descriptor.h"
+#include "entry-names.h"
+
+namespace Fortran::runtime {
+extern "C" {
+
+// Initializes the descriptor for an allocatable of intrinsic or derived type.
+// The incoming descriptor is treated as (and can be) uninitialized garbage.
+// Must be called for each allocatable variable as its scope comes into being.
+// The storage for the allocatable's descriptor must have already been
+// allocated to a size sufficient for the rank, corank, and type.
+// A descriptor must be initialized before being used for any purpose,
+// but needs reinitialization in a deallocated state only when there is
+// a change of type, rank, or corank.
+void RTNAME(AllocatableInitIntrinsic)(
+    Descriptor &, TypeCategory, int kind, int rank = 0, int corank = 0);
+void RTNAME(AllocatableInitCharacter)(Descriptor &, SubscriptValue length,
+    int kind = 1, int rank = 0, int corank = 0);
+void RTNAME(AllocatableInitDerived)(
+    Descriptor &, const DerivedType &, int rank = 0, int corank = 0);
+
+// For MOLD= and SOURCE= allocation with no (other) prior initialization.
+void RTNAME(AllocatableInitMold)(Descriptor &, const Descriptor &);
+
+// Explicitly sets the bounds and length type parameters of an initialized
+// deallocated allocatable.  (Not for use with MOLD= or SOURCE= in an
+// ALLOCATE statement; see below).  Length type parameters are indexed in
+// declaration order; i.e., 0 is the first length type parameter in the
+// deepest base type.
+void RTNAME(AllocatableSetBounds)(
+    Descriptor &, int dim, SubscriptValue lower, SubscriptValue upper);
+void RTNAME(AllocatableSetDerivedLength)(
+    Descriptor &, int which, SubscriptValue);
+
+// The upper bound is ignored for the last codimension.
+void RTNAME(AllocatableSetCoBounds)(
+    Descriptor &, int codim, SubscriptValue lower, SubscriptValue upper = 0);
```

Comment:
Are `codim` (and `dim` in `AllocatableSetBounds`) indexed starting from 1 as the `DIM` argument in Fortran intrinsic ?

Reply:
No, I intend for them to be zero-based, which is the convention used elsewhere in the runtime APIs for descriptors.  Would you prefer 1-based indices?

Reply:
No, zero based is fine (actually easier since FIR is already zero based here).